### PR TITLE
QoS 1 - Add interface for Subscription to register event callbacks

### PR DESCRIPTION
**Pull request by emersonknapp**
Depends on https://github.com/ros2/rclpy/pull/305
Adds just the new API for creating a subscription

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by nburek — APPROVED** [review]

File: rclpy/rclpy/node.py [hunk 1]
```diff
@@ -44,6 +44,7 @@
 from rclpy.qos import QoSProfile
 from rclpy.service import Service
 from rclpy.subscription import Subscription
+from rclpy.subscription import SubscriptionEventCallbacks
```
Comment (ross-desmond) [inline rclpy/rclpy/node.py:47]: Do we usually add a second line, or use a comma when importing a member from the same package?

File: rclpy/rclpy/subscription.py [hunk 2]
```diff
@@ -22,6 +22,26 @@
 MsgType = TypeVar('MsgType')
 
 
+class SubscriptionEventCallbacks:
+    """Container to provide middleware event callbacks for a Subscription."""
+
+    def __init__(
+        self,
+        deadline_callback: Callable = None,
```
Comment (ross-desmond) [inline rclpy/rclpy/subscription.py:30]: Python 3.7 `@dataclass` would be perfect here
https://docs.python.org/3/library/dataclasses.html

File: rclpy/rclpy/subscription.py [hunk 3]
```diff
@@ -52,8 +73,10 @@ def __init__(
         :param callback_group: The callback group for the subscription. If ``None``, then the
             nodes default callback group is used.
         :param qos_profile: The quality of service profile to apply to the subscription.
-        :node_handle: Capsule pointing to the underlying ``rcl_node_t`` object for the node the
-            subscription is associated with.
+        :param event_callbacks: Container for user-defined callback functions for middleware
+            events.
+        :param node_handle: Capsule pointing to the underlying ``rcl_node_t`` object for the node
```
Comment (ross-desmond) [inline rclpy/rclpy/subscription.py:78]: Nice comments

**Review by ross-desmond — APPROVED** [review]
Added minor enhancements, overall looks good

File: rclpy/rclpy/subscription.py [hunk 4]
```diff
@@ -33,6 +53,7 @@ def __init__(
          callback: Callable,
          callback_group: CallbackGroup,
          qos_profile: QoSProfile,
+         event_callbacks: SubscriptionEventCallbacks,
```
Comment (dabonnie) [inline rclpy/rclpy/subscription.py:56]: Curious why we're using a positional argument rather than optional?

File: rclpy/rclpy/node.py (hunk 1 above)
Comment (emersonknapp) [reply rclpy/rclpy/node.py:47]: Across this package it's one line per import, so i copied the style

File: rclpy/rclpy/subscription.py (hunk 2 above)
Comment (emersonknapp) [reply rclpy/rclpy/subscription.py:30]: I think we support Python3.5

File: rclpy/rclpy/subscription.py (hunk 4 above)
Comment (emersonknapp) [reply rclpy/rclpy/subscription.py:56]: This is an implementation API, the user-facing interface is `Node.create_subscription()`, which uses optional arguments

File: rclpy/rclpy/subscription.py (hunk 2 above)
Comment (emersonknapp) [reply rclpy/rclpy/subscription.py:30]: Yeah - found the official - Crystal suppored Python3.5 - I doubt they're bumping the version up for this release http://www.ros.org/reps/rep-2000.html#crystal-clemmys-december-2018-december-2019 - Ubuntu Xenial and Raspbian Stretch are both on 3.5 still

**Comment by emersonknapp** [comment]
The interfaces will actually look pretty different based on learnings in implementing - will reopen when ready

**Pull request closed without merging**